File: models/v1/fake/valid_ingredient.go
package fakemodels

import (
	models "gitlab.com/prixfixe/prixfixe/models/v1"

	fake "github.com/brianvoe/gofakeit/v5"
)

// BuildFakeValidIngredient builds a faked valid ingredient.
func BuildFakeValidIngredient() *models.ValidIngredient {
	return &models.ValidIngredient{
		ID:                 fake.Uint64(),
		Name:               fake.Word(),
		Variant:            fake.Word(),
		Description:        fake.Word(),
		Warning:            fake.Word(),
		ContainsEgg:        fake.Bool(),
		ContainsDairy:      fake.Bool(),
		ContainsPeanut:     fake.Bool(),
		ContainsTreeNut:    fake.Bool(),
		ContainsSoy:        fake.Bool(),
		ContainsWheat:      fake.Bool(),
		ContainsShellfish:  fake.Bool(),
		ContainsSesame:     fake.Bool(),
		ContainsFish:       fake.Bool(),
		ContainsGluten:     fake.Bool(),
		AnimalFlesh:        fake.Bool(),
		AnimalDerived:      fake.Bool(),
		MeasurableByVolume: fake.Bool(),
		Icon:               fake.Word(),
		CreatedOn:          uint64(uint32(fake.Date().Unix())),
	}
}

// BuildFakeValidIngredientList builds a faked ValidIngredientList.
func BuildFakeValidIngredientList() *models.ValidIngredientList {
	exampleValidIngredient1 := BuildFakeValidIngredient()
	exampleValidIngredient2 := BuildFakeValidIngredient()
	exampleValidIngredient3 := BuildFakeValidIngredient()

	return &models.ValidIngredientList{
		Pagination: models.Pagination{
			Page:       1,
			Limit:      20,
			TotalCount: 3,
		},
		ValidIngredients: []models.ValidIngredient{
			*exampleValidIngredient1,
			*exampleValidIngredient2,
			*exampleValidIngredient3,
		},
	}
}

// BuildFakeValidIngredientUpdateInputFromValidIngredient builds a faked ValidIngredientUpdateInput from a valid ingredient.
func BuildFakeValidIngredientUpdateInputFromValidIngredient(validIngredient *models.ValidIngredient) *models.ValidIngredientUpdateInput {
	return &models.ValidIngredientUpdateInput{
		Name:               validIngredient.Name,
		Variant:            validIngredient.Variant,
		Description:        validIngredient.Description,
		Warning:            validIngredient.Warning,
		ContainsEgg:        validIngredient.ContainsEgg,
		ContainsDairy:      validIngredient.ContainsDairy,
		ContainsPeanut:     validIngredient.ContainsPeanut,
		ContainsTreeNut:    validIngredient.ContainsTreeNut,
		ContainsSoy:        validIngredient.ContainsSoy,
		ContainsWheat:      validIngredient.ContainsWheat,
		ContainsShellfish:  validIngredient.ContainsShellfish,
		ContainsSesame:     validIngredient.ContainsSesame,
		ContainsFish:       validIngredient.ContainsFish,
		ContainsGluten:     validIngredient.ContainsGluten,
		AnimalFlesh:        validIngredient.AnimalFlesh,
		AnimalDerived:      validIngredient.AnimalDerived,
		MeasurableByVolume: validIngredient.MeasurableByVolume,
		Icon:               validIngredient.Icon,
	}
}

// BuildFakeValidIngredientCreationInput builds a faked ValidIngredientCreationInput.
func BuildFakeValidIngredientCreationInput() *models.ValidIngredientCreationInput {
	validIngredient := BuildFakeValidIngredient()
	return BuildFakeValidIngredientCreationInputFromValidIngredient(validIngredient)
}

// BuildFakeValidIngredientCreationInputFromValidIngredient builds a faked ValidIngredientCreationInput from a valid ingredient.
func BuildFakeValidIngredientCreationInputFromValidIngredient(validIngredient *models.ValidIngredient) *models.ValidIngredientCreationInput {
	return &models.ValidIngredientCreationInput{
		Name:               validIngredient.Name,
		Variant:            validIngredient.Variant,
		Description:        validIngredient.Description,
		Warning:            validIngredient.Warning,
		ContainsEgg:        validIngredient.ContainsEgg,
		ContainsDairy:      validIngredient.ContainsDairy,
		ContainsPeanut:     validIngredient.ContainsPeanut,
		ContainsTreeNut:    validIngredient.ContainsTreeNut,
		ContainsSoy:        validIngredient.ContainsSoy,
		ContainsWheat:      validIngredient.ContainsWheat,
		ContainsShellfish:  validIngredient.ContainsShellfish,
		ContainsSesame:     validIngredient.ContainsSesame,
		ContainsFish:       validIngredient.ContainsFish,
		ContainsGluten:     validIngredient.ContainsGluten,
		AnimalFlesh:        validIngredient.AnimalFlesh,
		AnimalDerived:      validIngredient.AnimalDerived,
		MeasurableByVolume: validIngredient.MeasurableByVolume,
		Icon:               validIngredient.Icon,
	}
}

File: frontend/src/models/validIngredient.ts
export class ValidIngredient {
     id: number;
     externalID: string;
     name: string;
     variant: string;
     description: string;
     warning: string;
     iconPath: string;
     containsSoy: boolean;
     containsTreeNut: boolean;
     containsShellfish: boolean;
     containsSesame: boolean;
     containsFish: boolean;
     containsGluten: boolean;
     animalFlesh: boolean;
     animalDerived: boolean;
     volumetric: boolean;
     containsPeanut: boolean;
     containsDairy: boolean;
     containsEgg: boolean;
     containsWheat: boolean;
     createdOn: number;
     lastUpdatedOn?: number;
     archivedOn?: number;

    constructor(
      id: number = 0,
      externalID: string = "",
      name: string = "",
      variant: string = "",
      description: string = "",
      warning: string = "",
      iconPath: string = "",
      containsSoy: boolean = false,
      containsTreeNut: boolean = false,
      containsShellfish: boolean = false,
      containsSesame: boolean = false,
      containsFish: boolean = false,
      containsGluten: boolean = false,
      animalFlesh: boolean = false,
      animalDerived: boolean = false,
      volumetric: boolean = false,
      containsPeanut: boolean = false,
      containsDairy: boolean = false,
      containsEgg: boolean = false,
      containsWheat: boolean = false,
      createdOn: number = 0,
      lastUpdatedOn?: number,
      archivedOn?: number,
    ) {
         this.id = id;
         this.externalID = externalID;
         this.name = name;
         this.variant = variant;
         this.description = description;
         this.warning = warning;
         this.iconPath = iconPath;
         this.containsSoy = containsSoy;
         this.containsTreeNut = containsTreeNut;
         this.containsShellfish = containsShellfish;
         this.containsSesame = containsSesame;
         this.containsFish = containsFish;
         this.containsGluten = containsGluten;
         this.animalFlesh = animalFlesh;
         this.animalDerived = animalDerived;
         this.volumetric = volumetric;
         this.containsPeanut = containsPeanut;
         this.containsDairy = containsDairy;
         this.containsEgg = containsEgg;
         this.containsWheat = containsWheat;
         this.createdOn = createdOn;
         this.lastUpdatedOn = lastUpdatedOn;
         this.archivedOn = archivedOn;
    }
}

export class ValidIngredientList {
     validIngredients: ValidIngredient[];
     totalCount: number;
     page: number;
     limit: number;
     filteredCount: number;

     constructor(
          validIngredients: ValidIngredient[] = [],
          totalCount: number = 0,
          page: number = 0,
          limit: number = 0,
          filteredCount: number = 0,
     ) {
          this.validIngredients = validIngredients;
          this.totalCount = totalCount;
          this.page = page;
          this.limit = limit;
          this.filteredCount = filteredCount;
     }
}